### PR TITLE
[22056] Transform locators using new machine_id PID

### PR DESCRIPTION
**Pull request by cferreiragonz**
## Description
This PR adds a new PID containing a UUID of the machine/host. It is used to check if the metadata of other participants comes from the same host in a more reliable approach than the GUID comparison, which uses IP addresses interfaces. It also checks if the locator used as key when a new channel is added to the `channel_resources_` map contains a local interface. In this case, a localhost locator is also linked to the same channel, to ensure that the participant can use a single channel for both localhost and the local interface. This does NOT duplicate sent messages, it only reuses the same TCP channel for two different locators, localhost and the specified IP address.

This feature solves a bug in which a TCP client connecting to `localhost` is unable to identify the locator of a TCP discovery server with a custom GUID listening on `any`. This causes the client to create two different channels for the same connection, where the latest can never be reached.  The reason is that the server's locator is not considered from the same host due to its custom GUID. Hence, the server transforms the client's locator into `localhost` but the client does not do the same with the server's locator.

@Mergifyio backport 3.1.x 3.0.x 2.14.x 2.10.x

## Contributor Checklist

- [X] Commit messages follow the project guidelines. 
- [X] The code follows the style guidelines of this project. 
- [X] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally. Related test: https://github.com/eProsima/Discovery-Server/pull/107
- [X] Any new/modified methods have been properly documented using Doxygen. 
- _N/A_ Any new configuration API has an equivalent XML API (with the corresponding XSD extension) 
- [X] Changes are backport compatible: they do **NOT** break ABI nor change library core behavior. 
- [X] Changes are API compatible. 
- [x] New feature has been added to the `versions.md` file (if applicable).
- _N/A_ New feature has been documented/Current behavior is correctly described in the documentation. 
    
- [X] Applicable backports have been included in the description.

## Reviewer Checklist

- [x] The PR has a milestone assigned.
- [x] The title and description correctly express the PR's purpose.
- [x] Check contributor checklist is correct.
- [ ] If this is a critical bug fix, backports to the critical-only supported branches have been requested.
- [ ] Check CI results: changes do not issue any warning.
- [ ] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

File: src/cpp/rtps/builtin/data/ParticipantProxyData.hpp [hunk 1]
```diff
@@ -77,6 +77,8 @@ class ParticipantProxyData
     ProtocolVersion_t m_protocolVersion;
     //!GUID
     GUID_t m_guid;
+    //!HOST ID
+    fastcdr::string_255 m_host_id;
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/data/ParticipantProxyData.hpp:81]:
```suggestion
    //!Machine ID
    fastcdr::string_255 machine_id;
```

File: src/cpp/rtps/builtin/data/ReaderProxyData.cpp [hunk 2]
```diff
@@ -191,6 +192,13 @@ uint32_t ReaderProxyData::get_serialized_size(
     // PID_NETWORK_CONFIGURATION_SET
     ret_val += 4 + PARAMETER_NETWORKCONFIGSET_LENGTH;
 
+    if (m_host_id.size() > 0)
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/rtps/builtin/data/ReaderProxyData.cpp:195]: Going one step further, we could avoid filtering the locators on `readFromCDRMessage`, and perform the filtering in a new method `setup_locators(const XXXProxyData& reader_data, NetworkFactory& network, const ParticipantProxyData& participant_data)` and call it [here](https://github.com/eProsima/Fast-DDS/blob/c6277440ccb6bae0e4533fd3a153752db88cac7d/src/cpp/rtps/builtin/discovery/endpoint/EDPSimpleListeners.cpp#L248-L260)

File: src/cpp/utils/SystemInfo.hpp [hunk 3]
```diff
@@ -77,6 +77,26 @@ class SystemInfo
...
+    /**
+     * Get the identifier of the current machine.
+     *
+     * @return the identifier of the current host.
+     */
+    inline fastcdr::string_255 check_machine_id() const
+    {
+        return Host::instance().machine_id();
+    }
+
```
Comment (MiguelCompany) [inline src/cpp/utils/SystemInfo.hpp:99]: Why two methods that do the same?

File: tools/fds/server.cpp [hunk 4]
```diff
@@ -569,6 +572,7 @@ int fastdds_discovery_server(
     }
 
     fastdds::rtps::GuidPrefix_t guid_prefix = participantQos.wire_protocol().prefix;
+    participantQos.transport().use_builtin_transports = true ? udp_server_initialized : false;
```
Comment (MiguelCompany) [inline tools/fds/server.cpp:575]: This is the same as `participantQos.transport().use_builtin_transports = udp_server_initialized;`. Please check it is really what you wanted.

File: src/cpp/rtps/builtin/data/ReaderProxyData.cpp [hunk 5]
```diff
@@ -1149,6 +1119,66 @@ bool ReaderProxyData::readFromCDRMessage(
...
+            }
+        }
+        auto locators = remote_locators_;
+        set_remote_locators(locators, network, true);
+    }
+    else
+    {
+        // Get the remote locators from the participant_data
+        set_remote_locators(participant_data.default_locators, network, true);
+    }
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/data/ReaderProxyData.cpp:1164]:
```suggestion
    if (rdata.has_locators())
    {
        // Get the transformed remote locators for the ReaderProxyData received
        remote_locators_.unicast.clear();
        remote_locators_.multicast.clear();
        for (const Locator_t& locator : rdata.remote_locators_.unicast)
        {
            Locator_t temp_locator;
            if (network.transform_remote_locator(locator, temp_locator, m_networkConfiguration, from_this_host))
            {
                ProxyDataFilters::filter_locators(network, remote_locators_, temp_locator, true);
            }
        }
        for (const Locator_t& locator : rdata.remote_locators_.multicast)
        {
            Locator_t temp_locator;
            if (network.transform_remote_locator(locator, temp_locator, m_networkConfiguration, from_this_host))
            {
                ProxyDataFilters::filter_locators(network, remote_locators_, temp_locator, false);
            }
        }
    }
    else
    {
        // Get the remote locators from the participant_data
        set_remote_locators(participant_data.default_locators, network, true, from_this_host);
    }
```

File: src/cpp/rtps/builtin/data/ReaderProxyData.cpp [hunk 6]
```diff
@@ -933,23 +891,7 @@ bool ReaderProxyData::readFromCDRMessage(
                             return false;
                         }
 
-                        if (!should_filter_locators)
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/rtps/builtin/data/ReaderProxyData.cpp:892]: Same for `WriterProxyData`

File: src/cpp/rtps/builtin/data/ReaderProxyData.cpp [hunk 7]
```diff
@@ -1149,6 +1119,66 @@ bool ReaderProxyData::readFromCDRMessage(
     return false;
 }
 
+void ReaderProxyData::setup_locators(
+        const ReaderProxyData* rdata,
+        NetworkFactory& network,
+        const ParticipantProxyData& participant_data)
+{
+    if (this == rdata)
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/rtps/builtin/data/ReaderProxyData.cpp:1127]: Same for `WriterProxyData`

File: src/cpp/rtps/builtin/data/ReaderProxyData.cpp [hunk 8]
```diff
@@ -1149,6 +1119,66 @@ bool ReaderProxyData::readFromCDRMessage(
...
+    {
+        return;
+    }
+
+    machine_id = participant_data.machine_id;
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/rtps/builtin/data/ReaderProxyData.cpp:1132]: Same for `WriterProxyData`

File: src/cpp/rtps/builtin/data/ReaderProxyData.cpp (hunk 5 above)
Comment (MiguelCompany) [reply src/cpp/rtps/builtin/data/ReaderProxyData.cpp:1164]: Same for `WriterProxyData`

File: src/cpp/rtps/builtin/data/ReaderProxyData.cpp [hunk 9]
```diff
@@ -1149,6 +1119,66 @@ bool ReaderProxyData::readFromCDRMessage(
...
+    {
+        same_host = machine_id == SystemInfo::instance().machine_id();
+    }
+    else
+    {
+        same_host = m_guid.is_from_this_host();
+    }
+    return same_host;
+}
+
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/data/ReaderProxyData.cpp:1181]:
```suggestion
```

Comment (MiguelCompany) [reply src/cpp/rtps/builtin/data/ReaderProxyData.cpp:1181]: Same for `WriterProxyData`

File: src/cpp/utils/Host.cpp [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+#include <IOKit/IOKitLib.h>
+#else
+#include <unistd.h>
+#include <fcntl.h>
+#endif // if defined(_WIN32)
+
+
+namespace eprosima {
+
+std::string Host::compute_machine_id()
```
Comment (MiguelCompany) [inline src/cpp/utils/Host.cpp:36]:
```suggestion
fastcdr::string_255 Host::compute_machine_id()
```

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

File: src/cpp/rtps/builtin/data/WriterProxyData.cpp [hunk 11]
```diff
@@ -1117,13 +1085,55 @@ bool WriterProxyData::readFromCDRMessage(
...
+    if (this == &wdata)
+    {
+        return;
+    }
+
+    bool from_this_host = participant_data.is_from_this_host();
+
+    if (wdata.has_locators())
+    {
+        // Get the transformed remote locators for the ReaderProxyData received
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/data/WriterProxyData.cpp:1102]:
```suggestion
        // Get the transformed remote locators for the WriterProxyData received
```

File: src/cpp/rtps/builtin/data/ReaderProxyData.hpp [hunk 12]
```diff
@@ -423,17 +427,26 @@ class ReaderProxyData
...
      */
     bool readFromCDRMessage(
             CDRMessage_t* msg,
-            NetworkFactory& network,
-            bool should_filter_locators,
             fastdds::rtps::VendorId_t source_vendor_id = c_VendorId_eProsima);
 
+    /**
+     * Transform and set the remote locators from the remote_locators_ of another ReaderProxyData.
+     * If the received WriterProxyData has no locators, remote locators will be extracted from the
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/data/ReaderProxyData.hpp:439]:
```suggestion
     * If the received ReaderProxyData has no locators, remote locators will be extracted from the
```

File: src/cpp/rtps/builtin/data/ReaderProxyData.hpp [hunk 13]
```diff
@@ -478,6 +491,8 @@ class ReaderProxyData
 
     //!GUID
     GUID_t m_guid;
+    //!HOST ID
+    fastcdr::string_255 machine_id;
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/data/ReaderProxyData.hpp:495]: Now we don't need this
```suggestion
```

File: src/cpp/rtps/builtin/data/WriterProxyData.hpp [hunk 14]
```diff
@@ -451,15 +455,29 @@ class WriterProxyData
...
+            const ParticipantProxyData& participant_data);
+
 private:
 
     //!GUID
     GUID_t m_guid;
 
+    //!HOST ID
+    fastcdr::string_255 machine_id;
+
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/data/WriterProxyData.hpp:480]:
```suggestion
```

File: src/cpp/rtps/builtin/discovery/participant/PDPClient.h [hunk 15]
```diff
@@ -169,15 +169,17 @@ class PDPClient : public PDP
...
 
     /**
      * Manually match the local PDP writer with the PDP reader of a given server. The function is
      * not thread safe (nts) in the sense that it does not take the PDP mutex. It does however take
      * temp_data_lock_
      */
     void match_pdp_reader_nts_(
-            const eprosima::fastdds::rtps::RemoteServerAttributes& server_att);
+            const eprosima::fastdds::rtps::RemoteServerAttributes& server_att,
+            bool from_this_host);
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/discovery/participant/PDPClient.h:182]: Update doxydoc on all modified methods

File: src/cpp/utils/Host.cpp [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+fastcdr::string_255 Host::compute_machine_id()
+{
+    #ifdef _WIN32
+    char machine_id[255];
+    DWORD BufferSize = sizeof(machine_id);
+    LONG res = RegGetValueA(HKEY_LOCAL_MACHINE, "SOFTWARE\\Microsoft\\Cryptography", "MachineGuid", RRF_RT_REG_SZ,
+                    NULL, machine_id, &BufferSize);
+    if (res == 0)
+    {
+        return std::string(machine_id);
```
Comment (MiguelCompany) [inline src/cpp/utils/Host.cpp:45]:
```suggestion
        return machine_id;
```

File: src/cpp/utils/Host.cpp [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+    char buffer[33] = {0};
+    ssize_t bytes_read = read(fd, buffer, 32);
+    close(fd);
+
+    if (bytes_read < 32)
+    {
+        return "";
+    }
+
+    return std::string(buffer, 32);
```
Comment (MiguelCompany) [inline src/cpp/utils/Host.cpp:85]:
```suggestion
    return fastcdr::string_255(buffer, 32);
```

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

File: tools/fds/server.cpp [hunk 18]
```diff
@@ -304,6 +304,8 @@ int fastdds_discovery_server(
     // Retrieve first TCP port
     option::Option* pO_tcp_port = options[TCP_PORT];
 
+    bool udp_server_initialized = true ? (pOp != nullptr || pO_port != nullptr) : false;
```
Comment (MiguelCompany) [inline tools/fds/server.cpp:307]:
```suggestion
    bool udp_server_initialized = (pOp != nullptr) || (pO_port != nullptr);
```

File: src/cpp/utils/Host.cpp [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+    char buffer[33] = {0};
+    ssize_t bytes_read = read(fd, buffer, 32);
+    close(fd);
+
+    if (bytes_read < 32)
+    {
+        return "";
+    }
+
+    return static_cast<fastcdr::string_255>(std::string(buffer, 32));
```
Comment (MiguelCompany) [inline src/cpp/utils/Host.cpp:85]:
```suggestion
    return fastcdr::string_255(buffer, 32);
```

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

**Review by MiguelCompany — DISMISSED** [review]
LGTM with green CI

File: test/unittest/transport/TCPv4Tests.cpp [hunk 20]
```diff
@@ -2032,7 +2035,7 @@ TEST_F(TCPv4Tests, client_announced_local_port_uniqueness)
 
     std::this_thread::sleep_for(std::chrono::milliseconds(100));
 
-    ASSERT_EQ(receiveTransportUnderTest.get_channel_resources().size(), 2u);
+    ASSERT_GT(receiveTransportUnderTest.get_channel_resources().size(), 2u);
```
Comment (MiguelCompany) [inline test/unittest/transport/TCPv4Tests.cpp:2038]: We could traverse the whole collection, add the values into a set, and check the size of the set to be equal to 2.

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

**Review by MiguelCompany — DISMISSED** [review]
LGTM with green CI

**Comment by cferreiragonz** [comment]
Shared memory is correctly enabled between docker containers.

**Pull request merged**